### PR TITLE
Sts008 receive notification letter

### PR DESCRIPTION
**Pull request by AdamHawtin**
# Motivation and Context
This changes adds the template for a social notification letter CSV template and the mapping for social action requests. This enables the action exporter service to generate the CSV files containing the addresses and other details required to generate the social notification letters. These CSV files are sent to an SFTP server to be picked up downstream.

# What has changed
* Added social notification file templates and template mapping rows to the SQL changes and updated changelogs accordingly.
* Added a docker-compose file to start up the services required for an integration test in docker locally. These run with names and ports specific to this service so should not conflict with the containers in [ras-rm-docker-dev](https://github.com/ONSdigital/ras-rm-docker-dev).
* Added an integration test which ensures that when an action instruction for a social notification is put on the relevant queue, this service picks it up, generates the correct CSV file and then sends it to the SFTP. 

# How to test?
Running a `mvn clean install` should now run the integration test for these changes, including running the required over services in docker and running this service locally.

# Links
Related to: https://github.com/ONSdigital/rm-action-service/pull/48
Trello card: [Dev task 5: INT: Notify respondent of invitation to a survey (5)](https://trello.com/c/xbFrDWWf/100-dev-task-5-int-notify-respondent-of-invitation-to-a-survey-5)

### CODE REVIEW

File: src/test/java/uk/gov/ons/ctp/response/action/export/service/TemplateServiceIT.java [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+
+        simpleMessageListener = new SimpleMessageListener(
+                rabbitConfig.getHost(),
+                rabbitConfig.getPort(),
+                rabbitConfig.getUsername(),
+                rabbitConfig.getPassword());
+    }
+
+    @Test
+    public void testTemplateGeneratesCorrectPrintFileForSocial() throws JAXBException, InterruptedException, IOException {
```
Comment (benjefferies) [inline src/test/java/uk/gov/ons/ctp/response/action/export/service/TemplateServiceIT.java]: @pricem14pc you might be interested to know there is now a test for this

File: src/test/java/uk/gov/ons/ctp/response/action/export/service/TemplateServiceIT.java [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+
+        // When
+        BlockingQueue<String> queue = simpleMessageListener.listen(
+                SimpleMessageBase.ExchangeType.Fanout,
+                "event-message-outbound-exchange");
+
+        queue.take();
+
+        // Then
+        final String sftpPath = "Documents/sftp/";
```
Comment (benjefferies) [inline src/test/java/uk/gov/ons/ctp/response/action/export/service/TemplateServiceIT.java]: Might be nice to refactor 104,110 out into a method called getLatestCsvFileOnSftp

File: src/test/java/uk/gov/ons/ctp/response/action/export/service/TemplateServiceIT.java [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+        assertEquals(actionRequest.getCaseRef(), templateRow.get(6));
+
+        defaultSftpSessionFactory.getSession().remove(notificationFilePath);
+
+    }
+
+    private CSVRecord readFirstCsvRow(InputStream inputStream) throws IOException {
+        Reader reader = new InputStreamReader(inputStream);
+        CSVParser parser = new CSVParser(reader, CSVFormat.newFormat(':'));
+        try {
```
Comment (benjefferies) [inline src/test/java/uk/gov/ons/ctp/response/action/export/service/TemplateServiceIT.java]: Might be able to use a try with resource here to tidy it up a bit

File: src/test/java/uk/gov/ons/ctp/response/action/export/service/TemplateServiceIT.java [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+
+        assertEquals(actionRequest.getAddress().getSampleUnitRef(), templateRow.get(0));
+        assertEquals(actionRequest.getAddress().getLine1(), templateRow.get(1));
+        assertEquals("", templateRow.get(2));
+        assertEquals(actionRequest.getAddress().getPostcode(), templateRow.get(3));
+        assertEquals(actionRequest.getAddress().getTownName(), templateRow.get(4));
+        assertEquals(actionRequest.getIac(), templateRow.get(5));
+        assertEquals(actionRequest.getCaseRef(), templateRow.get(6));
+
+        defaultSftpSessionFactory.getSession().remove(notificationFilePath);
```
Comment (benjefferies) [inline src/test/java/uk/gov/ons/ctp/response/action/export/service/TemplateServiceIT.java]: Maybe worth leaving a comment here to say we're deleting the file that was created

File: src/test/java/uk/gov/ons/ctp/response/action/export/service/TemplateServiceImplTest.java [hunk 5]
```diff
@@ -1,12 +1,13 @@
 package uk.gov.ons.ctp.response.action.export.service;
 
+import freemarker.template.Configuration;
```
Comment (benjefferies) [inline src/test/java/uk/gov/ons/ctp/response/action/export/service/TemplateServiceImplTest.java]: Imports can be tidied up

File: src/test/java/uk/gov/ons/ctp/response/action/export/service/TemplateServiceIT.java [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+
+    @Test
+    public void testTemplateGeneratesCorrectPrintFileForSocial() throws JAXBException, InterruptedException, IOException {
+        // Given
+        ActionRequest actionRequest = createSocialNotificationActionRequest();
+
+        ActionInstruction actionInstruction = new ActionInstruction();
+        actionInstruction.setActionRequest(actionRequest);
+
+        JAXBContext jaxbContext = JAXBContext.newInstance(ActionInstruction.class);
```
Comment (benjefferies) [inline src/test/java/uk/gov/ons/ctp/response/action/export/service/TemplateServiceIT.java]: Might be nice to refactor 87,89 out into a function like objectToXmlString

**Comment by AdamHawtin** [comment]
[Dev task 5: INT: Notify respondent of invitation to a survey (5)](https://trello.com/c/xbFrDWWf/100-dev-task-5-int-notify-respondent-of-invitation-to-a-survey-5)

File: src/main/resources/database/changes/release-10.50.0/add_caseref_to_print_template.sql [hunk 7]
```diff
@@ -0,0 +1,11 @@
+INSERT INTO actionexporter.template
```
Comment (JamesGardiner) [inline src/main/resources/database/changes/release-10.50.0/add_caseref_to_print_template.sql]: There is no county value here. There is in the letter template. It needs to be confirmed by the business that this won't break downstream before merging in.

Comment (benjefferies) [reply src/main/resources/database/changes/release-10.50.0/add_caseref_to_print_template.sql]: This has been added and confirmed by BA

**Pull request merged**